EIPS/eip-3690.md

```diff
@@ -0,0 +1,266 @@
+
+For "popular" contracts the gained efficiency is ~10x because the jumpdests section is relatively small compared to the code section and therefore there is much less bytes to loop over than in JUMPDEST analysis.
+
+## Reference Implementation
+
+We extend the `validate_opcode()` function of [EIP-3670](./eip-3670.md):
+```python
+# The same table as in EIP-3670
+valid_opcodes = ...
+
+# Remove JUMPDEST from the list of valid opcodes
+valid_opcodes.remove(0x5b)
+
+# This helper decodes a single unsigned LEB128 encoded value
+# This will abort on truncated (short) input
+def leb128u_decode(input: bytes) -> (int, int):
+  ret = 0
+  shift = 0
+  consumed_bytes = 0
+  while True:
+      # Check for truncated input
+      assert(consumed_bytes < len(input))
+      # Only allow up to 4-byte long leb128 encodings
+      assert(consumed_bytes <= 3)
+      input_byte = input[consumed_bytes]
+      consumed_bytes += 1
+      ret |= (input_byte & 0x7f) << shift
+      if (input_byte & 0x80) == 0:
+          # Do not allow additional leading zero bits.
+          assert(input_byte != 0 || consumed_bytes == 0)
+          break
+      shift += 7
+  return (ret, consumed_bytes)
+
+# This helper parses the jumpdest table into a list of relative offsets
+# This will abort on truncated (short) input
+def parse_table(input: bytes) -> list[int]:
+  jumpdests = []
+  pos = 0
+  while pos < len(input):
+      value, consumed_bytes = leb128u_decode(input[pos:])
+      jumpdests.append(value)
+      pos += consumed_bytes
+  return jumpdests
+
+# This helper translates the delta offsets into absolute ones
+# This will abort on invalid 0-value entries
+def process_jumpdests(delta: list[int]) -> list[int]:
+    jumpdests = []
+    partial_sum = 0
+    first = True
+    for d in delta:
+        if first:
+            first = False
+        else:
+            assert(d != 0)
+        partial_sum += d
+        jumpdests.append(CODE_SECTION_OFFSET + partial_sum)
+    return jumpdests
+
+# Fails with assertion on invalid code
+# Expects list of absolute jumpdest offsets
+def validate_code(code: bytes, jumpdests: list[int]):
+    pos = 0
+    while pos < len(code):
+        # Ensure the opcode is valid
+        opcode = code[pos]
+        pos += 1
+        assert(opcode in valid_opcodes)
+
+        # Remove touched offset
+        try:
+            jumpdests.remove(pos)
+        except ValueError:
+            pass
+
+        # Skip pushdata
+        if opcode >= 0x60 and opcode <= 0x7f:
+            pos += opcode - 0x60 + 1
+
+    # Ensure last PUSH doesn't go over code end
+    assert(pos == len(code))
+
+    # The table is invalid if there are untouched locations
+    assert(len(jumpdests) == 0)
+```
+
+## Test Cases
+
+#### Valid bytecodes
```

Comment:
```suggestion
### Valid bytecodes
```
Skipping section levels can have weird affects on tooling (like the ToC generator).

Reply:
Thanks, fixed.